core/src/test/scala/quasar/plugin/postgres/PostgresDestinationSpec.scala

```diff
@@ -170,50 +171,65 @@ object PostgresDestinationSpec extends EffectfulQSpec[IO] with CsvSupport with P
         }
     }
 
-    "upsert delete rows with string typed primary key" >>* {
```

Comment:
That we need to remove this smells a bit, and I think points to the need for a `Replace`/`Upsert` event as we're assuming that's what `Delete` means and that we can just ignore it if we're handling upserts elsewhere.

Alternatively, we could change our upsert implementation to filter on the _deleted_ ids (storing them in a separate temp table) rather than assuming we want to delete all the inserted ids.

Reply:
Right, if we would have semantical delete this would be much cleaner. In my opinion `DataEvent` isn't type safe enough, `Delete` is always being the first event is one example of it (as it's not encoded via types). Other examples
+ `Commit` _must_ be there, and it must be the last message
+ `Create` arg is `Chunk` and we _must_ handle everything using `Chunk`s not `Stream`s. 
+ There shouldn't be two subsequent `Commit`s. Current implementations work fine w/o this, but actually there is no sense in having two commits. 

> Alternatively, we could change our upsert implementation to filter on the deleted ids (storing them in a separate temp table) rather than assuming we want to delete all the inserted ids.

I don't like it 😐 This decouples actual data in tables and data to delete even more, but they're coupled. I mean that the very nature of upserting is that we grab data using offset and remove existing rows with ids presented in new iteration. :thinking: Doesn't this mean that `Delete` event is implementation leak?

Reply:
> This decouples actual data in tables and data to delete even more, but they're coupled. I mean that the very nature of upserting is that we grab data using offset and remove existing rows with ids presented in new iteration.

Only if you assume that's what the deletes mean, they're only coupled if you know why you have a data event stream, just looking at the stream doesn't couple them, which was the goal, but we're discovering other disadvantages.

Reply:
Yeah, :100: for more semantic flexibility. I wonder if it's possible to revert control, i.e. having a stream of whatever consumed by destination, periodically emit something that mimir/sdbe could use to determine last offset. This way we could fully move transaction/non-transaction to destination and have more options for retries and so on.